include/sys/arc_impl.h

```diff
@@ -169,6 +169,8 @@ typedef struct l1arc_buf_hdr {
 	/* self protecting */
 	refcount_t		b_refcnt;
 
+	/* protected by HDR_IO_IN_PROGRESS() */
+	zio_t			*b_zio;
```

Comment:
> In this specific case a better trade off is to probably add some lookup logic to find the relevant zio on the pending queue.

I'm not really sure if this is possible. @grwilson and I were talking about this a bit yesterday and the problem here is that raidz and gang blocks would make it really hard to just lookup zios in the queues.

> We want to keep the l1arc_buf_hdr as small as possible to maximize the useful ARC size. Adding a field for this particular rare case feels a bit extravagant.

I was thinking about this a little bit. I could put the pointer to the zio in the `arc_callback_t`, which is a much more transient struct. Would that make more sense? I'm also looking at the structs in `arc_impl.h` and we actually have a few opportunities there to compact `arc_buf_hdr_t` if that is important to us....

> It forces us to hold the hash_lock longer which in the past has been a point on contention.

So I will note that we are not holding the lock across any kind of IO (other than the one kmem_zalloc() which could theoretically require a swap operation). I don't think this will really have that bad of an impact on performance. The alternative would be to just retake the lock to assign the `zio` to the `hdr` and force the `HDR_IO_IN_PROGRESS()` code to loop back to `top` if it encountered a NULL pointer there (which would assume that the pointer would get filled in soon).

Reply:
> I'm not really sure if this is possible.

Indeed, that is an annoying complication.  Ahh well.

> I was thinking about this a little bit. I could put the pointer to the zio in the arc_callback_t, which is a much more transient struct. Would that make more sense? 

If possible I think it's worth keeping the `l1arc_buf_hdr` as small as possible.  It's not particularly uncommon for a system to have a few million of these allocated which does add up.

> have a few opportunities there to compact arc_buf_hdr_t if that is important to us....

If we can, we should.

> So I will note that we are not holding the lock across any kind of IO.

Yup, I agree it shouldn't be a big problem as long as the hash table is correctly sized but I did want to mention it as a potential concern.  I'm OK with the locking change, although let's avoid dropping and reacquiring it in `arc_read()`.  It a little uglier but better to drop it in the conditionals before `goto out;` then there's no need to reacquire.

Reply:
> If possible I think it's worth keeping the l1arc_buf_hdr as small as possible. It's not particularly uncommon for a system to have a few million of these allocated which does add up.

OK. I'll make the change.

> Yup, I agree it shouldn't be a big problem as long as the hash table is correctly sized but I did want to mention it as a potential concern. I'm OK with the locking change, although let's avoid dropping and reacquiring it in arc_read(). It a little uglier but better to drop it in the conditionals before goto out; then there's no need to reacquire.

I can't really do that without impacting performance because I need to drop the lock for the `zio_wait()` . This will only reacquire the lock in the case of a synchronous `arc_read()` where the L2ARC read failed.